### PR TITLE
:sparkles:Change mailbox controller to use space framework

### PR DESCRIPTION
**Pull request by andreyod**
## Summary
Change mailbox controller to use space framework and space-aware client instead of directly using workspaces.

## Related issue(s)
https://github.com/kubestellar/kubestellar/issues/853

Fixes #

### CODE REVIEW

**Review by effi-ofer — APPROVED** [review]
/LGTM

**Comment by clubanderson** [comment]
@andreyod @effi-ofer I assume this should no longer have 'do-not-merge' on it?  If not then I will remove my approval.  Let me know

**Comment by andreyod** [comment]
> @andreyod @effi-ofer I assume this should no longer have 'do-not-merge' on it? If not then I will remove my approval. Let me know

@clubanderson This is still WIP. Please remove your approval. I will remove 'do-not-merge' when it's ready.

File: cmd/mailbox-controller/controller.go [hunk 1]
```diff
@@ -220,33 +225,37 @@ func (ctl *mbCtl) sync(ctx context.Context, refany any) bool {
...
 		return true
 	}
 	// Now we have established that the SyncTarget exists and is not being deleted
 	if workspace == nil {
-		ws := &tenancyv1alpha1.Workspace{
+		//create space for mailbox cluster
+		space := &spacev1alpha1.Space{
 			ObjectMeta: metav1.ObjectMeta{
-				Annotations: map[string]string{SyncTargetNameAnnotationKey: syncTarget.Name},
+				Annotations: map[string]string{spacemanager.SpacePathAnnotationKey: ctl.espwPath},
```
Comment (MikeSpreitzer) [inline cmd/mailbox-controller/controller.go:241]: Why remove the sync target annotation?

File: cmd/mailbox-controller/main.go [hunk 2]
```diff
@@ -70,16 +74,20 @@ func main() {
...
+	//KubeStellar core space (KCS)
+	fs.StringVar(&kcsName, "kcs-name", kcsName, "the name of the KubeStellar core space")
+	fs.StringVar(&spaceProvider, "space-provider", spaceProvider, "the name of the KubeStellar space provider")
 
 	workloadClientOpts := clientopts.NewClientOpts("workload", "access to edge service provider workspace")
 	mbwsClientOpts := clientopts.NewClientOpts("mbws", "access to mailbox workspaces (really all clusters)")
 	mbwsClientOpts.SetDefaultCurrentContext("base")
 
-	inventoryClientOpts.AddFlags(fs)
+	spaceMgtClientOpts := clientopts.NewClientOpts("space-mgt", "access to space management workspaces")
```
Comment (MikeSpreitzer) [inline cmd/mailbox-controller/main.go:85]: "workspaces" plural?

**Comment by MikeSpreitzer** [comment]
The "edge.kcp.io" API group has been renamed to "edge.kubestellar.io".

File: cmd/mailbox-controller/main.go (hunk 2 above)
Comment (andreyod) [reply cmd/mailbox-controller/main.go]: Changed to singular.

File: cmd/mailbox-controller/controller.go (hunk 1 above)
Comment (andreyod) [reply cmd/mailbox-controller/controller.go]: Added back but note that it is set on Space object not a workspace.

**Comment by andreyod** [comment]
> The "edge.kcp.io" API group has been renamed to "edge.kubestellar.io".

Rebased

File: cmd/mailbox-controller/controller.go [hunk 3]
```diff
@@ -269,6 +281,8 @@ func (ctl *mbCtl) ensureEdgeBinding(ctx context.Context, workspace *tenancyv1alp
 		return true
 	}
 	logger = logger.WithValues("mbwsCluster", mbwsCluster, "bindingName", TheEdgeBindingName)
+	//TODO discuss: we can use space-aware client here, but because APIBinding is a third-party API,
+	//              we will need to call ConfigForSpace() and create clientset for each MBWS.
```
Comment (MikeSpreitzer) [inline cmd/mailbox-controller/controller.go:285]: Yes.  "third-party APIs" will be the norm.  We will eventually need to do this when we switch to using kube-bind.  We could use the space-aware client here right now, or wait 'till later.

File: cmd/mailbox-controller/main.go [hunk 4]
```diff
@@ -51,16 +51,20 @@ import (
...
 	edgeinformers "github.com/kubestellar/kubestellar/pkg/client/informers/externalversions"
+	"github.com/kubestellar/kubestellar/pkg/mcclient"
+	spacemanager "github.com/kubestellar/kubestellar/pkg/space-manager"
 )
 
 func main() {
 	resyncPeriod := time.Duration(0)
 	var concurrency int = 4
 	serverBindAddress := ":10203"
+	kcsName := "espw"
```
Comment (MikeSpreitzer) [inline cmd/mailbox-controller/main.go:66]: What makes this correspond with the kcp workspace path `root:espw` ?  That is, what makes a space-aware client that uses this space name actually talk to the `root:espw` workspace?

**Comment by MikeSpreitzer** [comment]
This PR needs to include an update of the doc for the mailbox controller, which is rendered at https://docs.kubestellar.io/main/Coding%20Milestones/PoC2023q1/mailbox-controller/

File: cmd/mailbox-controller/controller.go (hunk 1 above)
Comment (MikeSpreitzer) [reply cmd/mailbox-controller/controller.go]: Yes.
You raise the question of what user-controlled metadata is propagated from a space to the underlying provider object.  The answer is something that should be documented.

File: cmd/mailbox-controller/main.go [hunk 5]
```diff
@@ -70,16 +74,20 @@ func main() {
...
+	fs.StringVar(&kcsName, "kcs-name", kcsName, "the name of the KubeStellar core space")
+	fs.StringVar(&spaceProvider, "space-provider", spaceProvider, "the name of the KubeStellar space provider")
 
 	workloadClientOpts := clientopts.NewClientOpts("workload", "access to edge service provider workspace")
 	mbwsClientOpts := clientopts.NewClientOpts("mbws", "access to mailbox workspaces (really all clusters)")
 	mbwsClientOpts.SetDefaultCurrentContext("base")
 
-	inventoryClientOpts.AddFlags(fs)
+	spaceMgtClientOpts := clientopts.NewClientOpts("space-mgt", "access to space management workspace")
+	spaceMgtClientOpts.SetDefaultCurrentContext("root:space-mgt")
```
Comment (MikeSpreitzer) [inline cmd/mailbox-controller/main.go:86]: If this is a general convention then it should be documented.

**Pull request closed without merging**